File: server/src/main/scala/producer/message_generator/data_generators/relative_date_time/FixedRelativeDateTimeGenerator.scala
package producer.message_generator.data_generators.relative_date_time

import com.tools.teal.pulsar.ui.producer.v1.producer as pb
import producer.message_generator.json_generator.JsonGenerator

case class FixedRelativeDateTimeGenerator(
    value: Long,
    unit: DateTimeUnit,
    isRoundedToUnitStart: Boolean
)

object FixedRelativeDateTimeGenerator:
    def fromPb(v: pb.FixedRelativeDateTimeGenerator): FixedRelativeDateTimeGenerator =
        FixedRelativeDateTimeGenerator(
            value = v.value,
            unit = DateTimeUnit.fromPb(v.unit),
            isRoundedToUnitStart = v.isRoundedToUnitStart
        )

    def toPb(v: FixedRelativeDateTimeGenerator): pb.FixedRelativeDateTimeGenerator =
        pb.FixedRelativeDateTimeGenerator(
            value = v.value,
            unit = DateTimeUnit.toPb(v.unit),
            isRoundedToUnitStart = v.isRoundedToUnitStart
        )

File: ui/components/ui/ProducerSession/producer-task/message-generator/data-generators/relative-date-time/fixed-relative-date-time-generator.ts
import * as pb from '../../../../../../../grpc-web/tools/teal/pulsar/ui/producer/v1/producer_pb';
import { DateTimeUnit, dateTimeUnitFromPb, dateTimeUnitToPb } from './date-time-unit';

export type FixedRelativeDateTimeGenerator = {
  type: 'fixed-relative-date-time-generator',
  value: number,
  unit: DateTimeUnit,
  isRoundedToUnitStart: boolean
};

export function fixedRelativeDateTimeGeneratorFromPb(v: pb.FixedRelativeDateTimeGenerator): FixedRelativeDateTimeGenerator {
  return {
    type: 'fixed-relative-date-time-generator',
    value: v.getValue(),
    unit: dateTimeUnitFromPb(v.getUnit()),
    isRoundedToUnitStart: v.getIsRoundedToUnitStart()
  };
}

export function fixedRelativeDateTimeGeneratorToPb(v: FixedRelativeDateTimeGenerator): pb.FixedRelativeDateTimeGenerator {
  const message = new pb.FixedRelativeDateTimeGenerator();
  message.setValue(v.value);
  message.setUnit(dateTimeUnitToPb(v.unit));
  message.setIsRoundedToUnitStart(v.isRoundedToUnitStart);

  return message;
}
